### DISCUSSION TITLE
Way to use dot notation to refer to states in a state machine

### DISCUSSION THREAD

**Discussion by bgoolsby** [micropython/micropython#15694] [Using MicroPython] [answered]
I'm experienced writing C++ code in Arduino but am just starting in micropython.
I'm writing a simple state machine and would to define a type and variable something like this:
    type States = {idle, triggered, lockout}
    variable state : States
Then be able to use it in an If statement "if current_state == state.idle"
I know micropython doesn't support enums - is there another way to do it?  I like dot notation and would prefer that.
So far I'm just using strings, "if current_state == "idle"  but that seems prone to error.

**Comment by shariltumin** [comment, marked as answer]
You can use special methods of **dict** in Python to implement "dot notation". This simple module shows you how:

```py
# dotdict.py
_B='No such attribute: '
class Dot(dict):
    def __getattr__(B,N):
            if N in B:return B[N]
            else:raise AttributeError(_B+N)
    def __setattr__(B,N,V):
            B[N]=V
    def __delattr__(B,N):
            if N in B:del B[N]
            else:raise AttributeError(_B+N)
```

You can then use the class as shown below:

``` bash
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> from dotdict import Dot
>>> state=Dot()
>>> state.idle=1
>>> state.triggered=2
>>> state.lockout=3
>>> current_state=state.idle
>>> current_state
1
>>> state
{'triggered': 2, 'idle': 1, 'lockout': 3}
>>> if current_state==state.idle: print('Idle state')
... 
Idle state
>>> del state.stop
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "dotdict.py", line 14, in __delattr__
AttributeError: No such attribute: stop
>>> del state.triggered
>>> state
{'idle': 1, 'lockout': 3}
```

Hope this is of some use.

**Reply by bgoolsby** [reply]
Thanks very much - I'll try this

**Comment by IhorNehrutsa** [comment]
[docs/library/enum.rst: Add Enum class. #16842](https://github.com/micropython/micropython/pull/16842)

Inspired by @shariltumin `Dot class`.

**Comment by IhorNehrutsa** [comment]
Implementation in:
[micropython-lib/python-stdlib/enum/enum.py: Add Enum class. #980](https://github.com/micropython/micropython-lib/pull/980)
